proposals/0000-updated-partial-type-sigs.rst

```diff
@@ -0,0 +1,726 @@
+   What's awkward here is that the examples are given in order of increasing specificity; each example's type is more
+   specific than the previous. Yet GHC's behavior wibbles and wobbles between them.
+
+5. Sometimes, a type signature is meant to stand on its own; sometimes, it is meant to be understood in relation to
+   a nearby expression. For example, these type signatures stand on their own::
+
+     const :: a -> b -> a
+     prox :: Proxy a -> Proxy a
+     x :: Int
+
+   In all cases, we know everything there is to know about the type of these variables without looking further.
+   In the case of ``prox``, "knowing everything about the type" means that we can observe that ``a``\'s kind is
+   unconstrained. The full signature will be ``forall {k :: Type} (a :: k). Proxy @k a -> Proxy @k a``.
+
+   These type signatures do not stand on their own::
+
+     wurble :: _ -> Bool
+     wobble :: Either _ Bool -> Char
+
+   In order to know the types of ``wurble`` and ``wobble``, we must read their definitions.
+
+   Currently, we can distinguish between these cases by looking for the presence of an underscore (or, with
+   ``-XNamedWildCards``, a type variable spelled with an initial underscore) in the type signature. However,
+   sometimes we want a type signature with an underscore to stand alone (that is, be *complete*), and sometimes
+   we want a signature without an underscore not to (that is, be *partial*).
+
+   If we assume ``type Const :: forall a b. a -> b -> a``, then this type signature can be complete::
+
+     blurb :: Proxy (Const @_ @Bool 5) -> ()
+
+   We can infer that the first kind argument to ``Const`` should be ``Nat``. No definition for ``blurb``
+   is needed to know its type. Yet, our rule above that we distinguish complete from partial signatures
+   by the presence of an underscore works against us here.
+
+   Conversely, assuming ``f :: forall (a :: Bool). Proxy a -> ()``, this type signature can be partial::
+
+     hiccup :: Proxy a -> ()
+     hiccup = f
+
+   By looking at the *definition* for ``hiccup``, we learn that the kind of ``a`` should be ``Bool``. But
+   we cannot know that if the type signature is *complete*; indeed, GHC rejects ``hiccup`` today, believing
+   its type signature to be complete. The ``singletons`` package runs up against this: in the code it
+   generates, it needs the expression in order to infer the kind of a type variable. Yet, short of inserting
+   an underscore (and then asking our dear users to enable ``-XPartialTypeSignatures -Wno-partial-type-signatures``),
+   there is no way of getting the behavior we want.
+
+6. Control over partial type signatures vs complete type signatures is based on the presence or absence of
+   a wildcard. This means that GHC sometimes makes the wrong decision: it is conceivable to want a partial
+   type signature without a wildcard, and to write a complete signature with one.
+
+   Here is a desired-complete signature that has a missing piece::
```

Comment:
What does "desired-complete" mean? The user has specified an incomplete type signature here.